_posts/2018-01-09-ble-virtualization.md

```diff
@@ -0,0 +1,125 @@
+following application with different configurations:
+
+
+
+```c
+
+#include <simple_ble.h>
+#include <tock.h>
+
+// Sizes in bytes
+#define DEVICE_NAME_SIZE 7
+
+int main(void) {
+  // declarations of variables to be used in this BLE example application
+  uint16_t advertising_interval_ms = 100;
+  uint8_t device_name[] = "TockOS3";
+
+  // configure advertisement interval to 300ms
+  // configure LE only and discoverable
+  ble_initialize(advertising_interval_ms, true);
+
+  // configure device names as TockOS3
+  ble_advertise_name(device_name, DEVICE_NAME_SIZE);
+
+  // start advertising
+  ble_start_advertising();
+
+  return 0;
+}
+
+```
+
+While running a Bluetooth Sniffer we find the four different Tock devices
+with different configurations as the screenshot below shows:
+
+
+![Bluetooth Low Energy](../assets/2018/01/ble_advertising.png)
+
+
+Note, that the kernel manages to generate unique advertisement addresses known
+as AdvA in the Bluetooth standard. Also, you can see that the applications by
+themselves configures different advertisement intervals and device names.
+TockOS2 is actually configured as Thermometer service with dummy data.
+
+
+## Implementation details
+First of all, we have not blogged about how the Bluetooth Low Energy driver works.
+The functionality is limited and support only advertisements and passive scanning so
+far i.e., no connections, different security modes and similar.
+
+
+Overall, the driver works in the following sequence:
+1. Initialize different Bluetooth parameters
+(advertisement interval, device name, other gap data and etc)
+2. Start advertise or start scanning, (this implies to configure virtual timer alarms)
+3. The virtual alarm expires and the radio gets activated for short while (magnitude in terms of milliseconds), for example by sending advertisements
+4. Reconfigure virtual alarm and go to sleep
+5. Step 3 and 4 continues forever until an application explicitly stops the operation
+
+
+However, in this blog post we want explain how we can use Tock architecture to
+achieve high degree of concurrency in the Bluetooth Low Energy driver and
+the different types of trade-offs we have performed with the implementation.
+A fundamental design decision of Tock is that the kernel is statically allocated
+which implies that the kernel can't dynamically allocate. Instead Tock uses an
+unique process model with five segments which the figure below shows:
+
+![Tock architecture](../assets/img/architecture.png)
+
+In context of this driver the grant segment is the essential part which makes
+it possible for the kernel to allocate memory and store data for each process.
+The grant contains Bluetooth Low Energy related data such as transmitting
+power, advertising interval, the actual data to advertise, and state along with
+virtual timer alarm.
+
+
+First, let's discuss the virtual alarm because it essential
+for correctness that the correct process is identified. Because, the virtual
```

Comment:
"right process"

It's weird to read "for correctness" and then "correct process" IMO